File: excel_generator.py
"""Генератор Excel-файлов с извещениями на оплату."""

from openpyxl import Workbook
from openpyxl.styles import Font, Alignment, PatternFill, Border, Side
from models import PaymentNotice
from date_time_utils import DateTimeHandler


class ExcelGenerator:
    """Генератор Excel-файлов."""

    def __init__(self):
        self._date_handler = DateTimeHandler()

    def generate_payment_notice(self, notice: PaymentNotice, file_path: str) -> None:
        """Генерирует Excel-файл с извещением на оплату."""
        wb = Workbook()
        ws = wb.active
        ws.title = "Извещение на оплату"

        # Стили
        header_font = Font(bold=True, size=14, color="FFFFFF")
        header_fill = PatternFill(start_color="366092", end_color="366092", fill_type="solid")
        label_font = Font(bold=True)
        border = Border(
            left=Side(style='thin'), right=Side(style='thin'),
            top=Side(style='thin'), bottom=Side(style='thin')
        )

        # Заголовок
        ws.merge_cells('A1:E1')
        ws['A1'].value = "ИЗВЕЩЕНИЕ НА ОПЛАТУ"
        ws['A1'].font = header_font
        ws['A1'].fill = header_fill
        ws['A1'].alignment = Alignment(horizontal='center', vertical='center')

        # Информация о лицевом счете
        row = 3
        for label, value in [
            ("Лицевой счет:", notice.account.account_number),
            ("ФИО:", notice.account.full_name),
            ("Адрес:", notice.account.get_address(notice.street)),
            ("Период:", f"{self._date_handler.get_month_name(notice.period_month)} {notice.period_year}"),
            ("Дата формирования:", self._date_handler.format_date(self._date_handler.get_current_date()))
        ]:
            ws[f'A{row}'] = label
            ws[f'B{row}'] = value
            ws[f'A{row}'].font = label_font
            row += 1

        # Таблица начислений
        row += 1
        headers = ["№", "Услуга", "Количество", "Тариф", "Сумма"]
        for col, header in enumerate(headers, start=1):
            cell = ws.cell(row=row, column=col)
            cell.value = header
            cell.font = header_font
            cell.fill = header_fill
            cell.alignment = Alignment(horizontal='center', vertical='center')
            cell.border = border

        # Данные начислений
        row += 1
        for idx, (charge, service) in enumerate(notice.charges, start=1):
            ws.cell(row=row, column=1, value=idx).border = border
            ws.cell(row=row, column=1).alignment = Alignment(horizontal='center')
            ws.cell(row=row, column=2, value=service.name).border = border
            ws.cell(row=row, column=3, value=charge.quantity).border = border
            ws.cell(row=row, column=3).alignment = Alignment(horizontal='right')
            ws.cell(row=row, column=4, value=service.tariff).border = border
            ws.cell(row=row, column=4).alignment = Alignment(horizontal='right')
            cost = service.calculate_cost(charge.quantity)
            ws.cell(row=row, column=5, value=cost).border = border
            ws.cell(row=row, column=5).alignment = Alignment(horizontal='right')
            row += 1

        # Итого
        ws.merge_cells(f'A{row}:D{row}')
        ws[f'A{row}'].value = "ИТОГО К ОПЛАТЕ:"
        ws[f'A{row}'].font = Font(bold=True, size=12)
        ws[f'A{row}'].alignment = Alignment(horizontal='right', vertical='center')
        ws[f'A{row}'].border = border
        ws[f'E{row}'].value = notice.total_amount
        ws[f'E{row}'].font = Font(bold=True, size=12)
        ws[f'E{row}'].alignment = Alignment(horizontal='right', vertical='center')
        ws[f'E{row}'].border = border
        ws[f'E{row}'].fill = PatternFill(start_color="FFFF00", end_color="FFFF00", fill_type="solid")

        # Ширина столбцов
        ws.column_dimensions['A'].width = 8
        ws.column_dimensions['B'].width = 40
        ws.column_dimensions['C'].width = 15
        ws.column_dimensions['D'].width = 15
        ws.column_dimensions['E'].width = 15

        wb.save(file_path)

File: js_project/excelGenerator.js
/**
 * Генератор Excel-файлов с извещениями на оплату.
 */

import ExcelJS from 'exceljs';
import { DateTimeHandler } from './dateTimeUtils.js';

/**
 * Класс генератора Excel-файлов.
 * 
 * Демонстрирует:
 * 1. ИНКАПСУЛЯЦИЮ: логика генерации Excel скрыта внутри класса
 * 2. КОМПОЗИЦИЮ: использует DateTimeHandler для работы с датами
 * 3. АБСТРАКЦИЮ: предоставляет простой метод generatePaymentNotice(),
 *    скрывающий сложность работы с библиотекой ExcelJS
 */
export class ExcelGenerator {
    // Поле для хранения ссылки на класс DateTimeHandler
    #dateHandler;

    constructor() {
        // Инициализация: сохраняем ссылку на класс DateTimeHandler
        // DateTimeHandler - это класс со статическими методами
        this.#dateHandler = DateTimeHandler;
    }

    /**
     * Генерирует Excel-файл с извещением на оплату.
     * @param {PaymentNotice} notice - Извещение на оплату
     * @param {string} filePath - Путь к файлу
     * @returns {Promise<void>}
     */
    async generatePaymentNotice(notice, filePath) {
        const workbook = new ExcelJS.Workbook();
        const worksheet = workbook.addWorksheet('Извещение на оплату');

        // Стили
        const headerFont = {
            name: 'Arial',
            size: 14,
            bold: true,
            color: { argb: 'FFFFFFFF' }
        };
        const headerFill = {
            type: 'pattern',
            pattern: 'solid',
            fgColor: { argb: 'FF366092' }
        };
        const labelFont = {
            name: 'Arial',
            bold: true
        };
        const border = {
            top: { style: 'thin' },
            left: { style: 'thin' },
            bottom: { style: 'thin' },
            right: { style: 'thin' }
        };

        // Заголовок
        worksheet.mergeCells('A1:E1');
        const headerCell = worksheet.getCell('A1');
        headerCell.value = 'ИЗВЕЩЕНИЕ НА ОПЛАТУ';
        headerCell.font = headerFont;
        headerCell.fill = headerFill;
        headerCell.alignment = { horizontal: 'center', vertical: 'middle' };
        headerCell.border = border;

        // Информация о лицевом счете
        let row = 3;
        const info = [
            ['Лицевой счет:', notice.account.accountNumber],
            ['ФИО:', notice.account.fullName],
            ['Адрес:', notice.account.getAddress(notice.street)],
            ['Период:', `${this.#dateHandler.getMonthName(notice.periodMonth)} ${notice.periodYear}`],
            ['Дата формирования:', this.#dateHandler.formatDate(this.#dateHandler.getCurrentDate())]
        ];

        for (const [label, value] of info) {
            const labelCell = worksheet.getCell(`A${row}`);
            labelCell.value = label;
            labelCell.font = labelFont;
            
            const valueCell = worksheet.getCell(`B${row}`);
            valueCell.value = value;
            row++;
        }

        // Таблица начислений
        row++;
        const headers = ['№', 'Услуга', 'Количество', 'Тариф', 'Сумма'];
        for (let col = 1; col <= headers.length; col++) {
            const cell = worksheet.getCell(row, col);
            cell.value = headers[col - 1];
            cell.font = headerFont;
            cell.fill = headerFill;
            cell.alignment = { horizontal: 'center', vertical: 'middle' };
            cell.border = border;
        }

        // Данные начислений
        row++;
        for (let idx = 0; idx < notice.charges.length; idx++) {
            const [charge, service] = notice.charges[idx];
            const cost = service.calculateCost(charge.quantity);

            // №
            const numCell = worksheet.getCell(row, 1);
            numCell.value = idx + 1;
            numCell.border = border;
            numCell.alignment = { horizontal: 'center' };

            // Услуга
            const serviceCell = worksheet.getCell(row, 2);
            serviceCell.value = service.name;
            serviceCell.border = border;

            // Количество
            const quantityCell = worksheet.getCell(row, 3);
            quantityCell.value = charge.quantity;
            quantityCell.border = border;
            quantityCell.alignment = { horizontal: 'right' };

            // Тариф
            const tariffCell = worksheet.getCell(row, 4);
            tariffCell.value = service.tariff;
            tariffCell.border = border;
            tariffCell.alignment = { horizontal: 'right' };

            // Сумма
            const costCell = worksheet.getCell(row, 5);
            costCell.value = cost;
            costCell.border = border;
            costCell.alignment = { horizontal: 'right' };

            row++;
        }

        // Итого
        worksheet.mergeCells(`A${row}:D${row}`);
        const totalLabelCell = worksheet.getCell(`A${row}`);
        totalLabelCell.value = 'ИТОГО К ОПЛАТЕ:';
        totalLabelCell.font = { name: 'Arial', size: 12, bold: true };
        totalLabelCell.alignment = { horizontal: 'right', vertical: 'middle' };
        totalLabelCell.border = border;

        const totalAmountCell = worksheet.getCell(`E${row}`);
        totalAmountCell.value = notice.totalAmount;
        totalAmountCell.font = { name: 'Arial', size: 12, bold: true };
        totalAmountCell.alignment = { horizontal: 'right', vertical: 'middle' };
        totalAmountCell.border = border;
        totalAmountCell.fill = {
            type: 'pattern',
            pattern: 'solid',
            fgColor: { argb: 'FFFFFF00' }
        };

        // Ширина столбцов
        worksheet.getColumn(1).width = 8;
        worksheet.getColumn(2).width = 40;
        worksheet.getColumn(3).width = 15;
        worksheet.getColumn(4).width = 15;
        worksheet.getColumn(5).width = 15;

        await workbook.xlsx.writeFile(filePath);
    }
}

